### PR TITLE
[Subtask]: Improve Operator Communication in Pipeline

### PR DESCRIPTION
**Pull request by badboynt1**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #11824

## What this PR does / why we need it:
implement a pipeline message framework.
optimize scan->top

### CODE REVIEW

File: pkg/sql/colexec/table_scan/table_scan.go [hunk 1]
```diff
@@ -63,6 +66,17 @@ func (arg *Argument) Call(proc *process.Process) (vm.CallResult, error) {
...
+		if arg.msgReceiver != nil {
+			msgs := arg.msgReceiver.ReceiveMessage(false)
+			for i := range msgs {
+				msg, ok := msgs[i].(process.TopValueMessage)
+				if !ok {
+					panic("only support top value message in table scan!")
+				}
+				arg.Reader.SetFilterZM(msg.TopValueZM)
+			}
+		}
```
Comment (ouyuanning) [inline pkg/sql/colexec/table_scan/table_scan.go:79]: Analyze需要剔除这部分等待的时间。

Comment (badboynt1) [reply pkg/sql/colexec/table_scan/table_scan.go:79]: 这个message是非block的，不需要wait。应该不影响analyze吧

Comment (ouyuanning) [reply pkg/sql/colexec/table_scan/table_scan.go:79]: arg.msgReceiver.ReceiveMessage 不是阻塞的话，那应该没影响

**Review by fengttt — COMMENTED** [review]
I have a overall comment
- for top, there should be a test -- actually there should be a config (session variable, or query hint) so that we can control the top push down.   then a test case compare different plan and explain analyze.
- what is the sql that the top can be applied?    can top be pushed down across aggregate, join, etc etc?

File: pkg/sql/colexec/top/top.go [hunk 2]
```diff
@@ -253,3 +265,90 @@ func (ctr *container) sort() {
 	}
 	heap.Init(ctr)
 }
+
+func (arg *Argument) updateTopValueZM() bool {
```
Comment (fengttt) [inline pkg/sql/colexec/top/top.go:269]: can you get message working for multi col?    select * from T order by a, b, c limit 5?

**Review by fengttt — COMMENTED** [review]
It maybe better to split the PR into 2, one for message mechanism, the other for top message implementation.   So that it can serve as an example for people who want to add a new operator that can handle a message.

**Comment by badboynt1** [comment]
> It maybe better to split the PR into 2, one for message mechanism, the other for top message implementation. So that it can serve as an example for people who want to add a new operator that can handle a message.

其他人想要实现message的话，参考这个就可以了
[topValueMsg.go](https://github.com/matrixorigin/matrixone/pull/14568/files#diff-c76764f05084b462f4f05470c73831626aef2e7212c7707da58d8693baeff225)

File: pkg/sql/colexec/top/top.go (hunk 2 above)
Comment (badboynt1) [reply pkg/sql/colexec/top/top.go:269]: 目前只支持单列。需要支持多列的话，也留出了接口，可以等以后的pr再实现。

**Comment by badboynt1** [comment]
> I have a overall comment
> 
> * for top, there should be a test -- actually there should be a config (session variable, or query hint) so that we can control the top push down.   then a test case compare different plan and explain analyze.
> * what is the sql that the top can be applied?    can top be pushed down across aggregate, join, etc etc?

如果只是为了测试不同的plan，需要以后实现hint机制。有另外的issue来跟踪hint机制。

top发送message到scan，目前可以越过inner和semi join。 left join会将limit算子直接下推，不需要在这里处理。 agg是不能越过的，没有意义

**Review by ouyuanning — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Comment by nnsgmsone** [comment]
那个top的例子加个bvt吧

**Review by heni02 — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Pull request merged**